packages/shared/ReactComponentStackFrame.js

```diff
@@ -76,89 +87,148 @@ export function describeNativeComponentFrame(
     }
   }
 
-  let control;
-
   reentry = true;
   const previousPrepareStackTrace = Error.prepareStackTrace;
   // $FlowFixMe[incompatible-type] It does accept undefined.
   Error.prepareStackTrace = undefined;
   let previousDispatcher;
+
   if (__DEV__) {
     previousDispatcher = ReactCurrentDispatcher.current;
     // Set the dispatcher in DEV because this might be call in the render function
     // for warnings.
     ReactCurrentDispatcher.current = null;
     disableLogs();
   }
-  try {
-    // This should throw.
-    if (construct) {
-      // Something should be setting the props in the constructor.
-      const Fake = function () {
-        throw Error();
-      };
-      // $FlowFixMe[prop-missing]
-      Object.defineProperty(Fake.prototype, 'props', {
-        set: function () {
-          // We use a throwing setter instead of frozen or non-writable props
-          // because that won't throw in a non-strict mode function.
-          throw Error();
-        },
-      });
-      if (typeof Reflect === 'object' && Reflect.construct) {
-        // We construct a different control for this case to include any extra
-        // frames added by the construct call.
-        try {
-          Reflect.construct(Fake, []);
-        } catch (x) {
-          control = x;
-        }
-        Reflect.construct(fn, [], Fake);
-      } else {
-        try {
-          Fake.call();
-        } catch (x) {
-          control = x;
-        }
-        // $FlowFixMe[prop-missing] found when upgrading Flow
-        fn.call(Fake.prototype);
+
+  /**
+   * Finding a common stack frame between sample and control errors can be
+   * tricky given the different types and levels of stack trace truncation from
+   * different JS VMs. So instead we'll attempt to control what that common
+   * frame should be through this class:
+   * Having both the sample and control errors be under the
+   * `DescribeNativeComponentFrameRoot` method call will ensure that a stack
+   * frame exists that has the method name `DescribeNativeComponentFrameRoot` in
+   * it for both control and sample stacks.
+   *
+   * Note that we're using a class method here instead of a plain object
+   * property to prevent Closure compiler from eliding away the object and the
+   * extra method call.
+   */
+  class RunInRootFrame {
+    constructor() {
+      // Bun and Safari will require setting these properties should the class
+      // ever get transpiled into an ES2015 constructor function.
+      // $FlowFixMe[method-unbinding]
+      this.DetermineComponentFrameRoot.displayName =
+        'DetermineComponentFrameRoot';
+
+      // Before ES6, the `name` property was not configurable.
+      if (
+        // $FlowFixMe[method-unbinding]
+        Object.getOwnPropertyDescriptor(this.DetermineComponentFrameRoot)
```

Comment:
Hmm should this be `Object.getOwnPropertyDescriptor(this.DetermineComponentFrameRoot, 'name')`?

Also JS q -- if the `name` property is not configurable, does that mean it should already hold the value `DetermineComponentFrameRoot`?

Reply:
Oh yup, I should specify `'name'` there!
>  the `name` property is not configurable, does that mean it should already hold the value `DetermineComponentFrameRoot`

I believe it depends on the JS VM 😕. For example, it's possible for:
```
RunInRootFrame.prototype.DetermineComponentFrameRoot =
  function DetermineComponentFrameRoot() {
    ...
  };
```
to get minified to:
```
a.prototype.DetermineComponentFrameRoot = function b() {
  ...
};
```

In this case, I think most VMs would set the `name` property here to `"b"`. Pre-ES6 it wouldn't be possible to change that, ES6 onwards, the name property is not writable (so we can't directly set it via property access), but is configurable, so we can change it via `Object.defineProperty`. Also some VMs might specify both the function name *and* method name in stack traces, so e.g. V8 would have a stack trace line like:
```
    at a.b [as DetermineComponentFrameRoot] 
```